### PR TITLE
feat: 🎸 add "truncated" field to /first-rows

### LINKED ISSUE #1731: add `truncated` field to /first-rows
On the Hub's side, we have to guess if the result has been truncated. For example, if the result has 35 rows: maybe it has not been truncated because the split only had 35 rows, or maybe it has been truncated because the cells contain heavy data -> *we cannot detect*.

It would be more explicit to return `truncated: boolean` in the response.

### PR DESCRIPTION
**Pull request by severo**
fixes #1731.

It will require refreshing split-first-rows-from-streaming and split-first-rows-from-parquet for all the datasets. The backfill will do it since I updated the versions.

Also, I put the "truncated" field as optional for now, but once the update is complete, I'll create another PR to make it mandatory (we cannot do a database migration since we don't know the value).

### CODE REVIEW

File: services/worker/src/worker/job_runners/split/first_rows_from_parquet.py [hunk 1]
```diff
@@ -103,7 +104,10 @@ def compute_first_rows_response(
 
     # get the rows
     try:
-        pa_table = rows_index.query(offset=0, length=rows_max_number)
+        pa_table_plus_one = rows_index.query(offset=0, length=rows_max_number + 1)
+        # ^^ to be able to detect if a split has exactly rows_max_number rows
+        all_fetched = len(pa_table_plus_one) <= rows_max_number
+        pa_table = pa_table_plus_one.slice(offset=0, length=rows_max_number)
```
Comment (lhoestq) [inline services/worker/src/worker/job_runners/split/first_rows_from_parquet.py:110]: simpler:

```suggestion
        pa_table = rows_index.query(offset=0, length=rows_max_number)
        all_fetched = rows_index.parquet_index.num_rows_total <= rows_max_number
```

File: services/worker/src/worker/job_runners/split/image_url_columns.py [hunk 2]
```diff
@@ -62,6 +62,7 @@ def compute_image_url_columns(
             split=split,
             features=first_rows_response["content"]["features"],
             rows=first_rows_response["content"]["rows"],
+            truncated=first_rows_response["content"]["truncated"] if "truncated" in first_rows_response else None,
```
Comment (lhoestq) [inline services/worker/src/worker/job_runners/split/image_url_columns.py:65]:
```suggestion
            truncated=first_rows_response["content"]["truncated"] if "truncated" in first_rows_response["content"] else None,
```

**Review by lhoestq — APPROVED** [review]
Maybe only increment the version of the parquet one first and then the streaming one ?
This way we won't have the risk of having concurrency issues.

Also note that this will also trigger the child jobs, e.g opt-in-out-url-scan

**Comment by severo** [comment]
> This way we won't have the risk of having concurrency issues.

besides possible concurrency issues, it would be a way to spend less on computing. Good idea.

**Comment by severo** [comment]
Followup PRs:
- increase version of `split-first-rows-from-streaming`
- set `truncated` field as mandatory, not optional

**Comment by severo** [comment]
As of today:

<details><summary>list of remaining cache entries</summary>
<pre><code>
db.cachedResponsesBlue.aggregate([
  {$match: {kind: "split-first-rows-from-parquet", http_status: 200, "content.truncated": {"$exists": false}}},
  {$group: {_id: "$dataset", count: {$sum: 1}}},
  {$sort: {count: -1}},
  {$project: {dataset: "$_id", _id: 0, count: 1}}
])
</code></pre>
<pre><code>
{ count: 21, dataset: 'Muennighoff/multi_eurlex' }
{ count: 8, dataset: 'bigbio/cellfinder' }
{ count: 3, dataset: 'google/xtreme_s' }
{ count: 2, dataset: 'siddharthtumre/Revised-JNLPBA' }
{ count: 1, dataset: 'Maxssto/mysetmp' }
{ count: 1, dataset: 'suryajg/suraj-1234' }
{ count: 1, dataset: 'MANMEET75/emotions' }
{ count: 1, dataset: 'distil-whisper/librispeech_asr-prompted' }
{ count: 1,
  dataset: 'thanhduycao/data_for_synthesis_with_entities_align_v4' }
{ count: 1, dataset: 'dabdib/collections' }
{ count: 1, dataset: 'nc33/CLM_data' }
{ count: 1, dataset: 'mozilla-foundation/common_voice_11_0' }
</code></pre>


<pre><code>
db.cachedResponsesBlue.aggregate([
  {$match: {kind: "split-first-rows-from-streaming", http_status: 200, "content.truncated": {"$exists": false}}},
  {$group: {_id: "$dataset", count: {$sum: 1}}},
  {$sort: {count: -1}},
  {$project: {dataset: "$_id", _id: 0, count: 1}}
])
</code></pre>
<pre><code>
{ count: 445,
  dataset: 'open-llm-leaderboard/details_meta-llama__Llama-2-7b-hf' }
{ count: 435,
  dataset: 'open-llm-leaderboard/details_meta-llama__Llama-2-70b-hf' }
{ count: 417,
  dataset: 'open-llm-leaderboard/details_openlm-research__open_llama_3b' }
{ count: 369,
  dataset: 'open-llm-leaderboard/details_meta-llama__Llama-2-13b-hf' }
{ count: 307,
  dataset: 'open-llm-leaderboard/details_togethercomputer__RedPajama-INCITE-7B-Base' }
{ count: 307,
  dataset: 'open-llm-leaderboard/details_facebook__opt-66b' }
{ count: 306,
  dataset: 'open-llm-leaderboard/details_huggyllama__llama-30b' }
{ count: 301,
  dataset: 'open-llm-leaderboard/details_openlm-research__open_llama_7b_v2' }
{ count: 300,
  dataset: 'open-llm-leaderboard/details_huggyllama__llama-7b' }
{ count: 300,
  dataset: 'open-llm-leaderboard/details_openlm-research__open_llama_13b' }
{ count: 297,
  dataset: 'open-llm-leaderboard/details_bigcode__starcoder' }
{ count: 253,
  dataset: 'open-llm-leaderboard/details_Open-Orca__OpenOrcaxOpenChat-Preview2-13B' }
{ count: 252,
  dataset: 'open-llm-leaderboard/details_tiiuae__falcon-40b' }
{ count: 248,
  dataset: 'open-llm-leaderboard/details_TheBloke__Llama-2-13B-GPTQ' }
{ count: 246,
  dataset: 'open-llm-leaderboard/details_OpenAssistant__codellama-13b-oasst-sft-v10' }
{ count: 245,
  dataset: 'open-llm-leaderboard/details_Salesforce__codegen-16B-nl' }
{ count: 242,
  dataset: 'open-llm-leaderboard/details_facebook__opt-30b' }
{ count: 242,
  dataset: 'open-llm-leaderboard/details_tiiuae__falcon-7b' }
{ count: 239,
  dataset: 'open-llm-leaderboard/details_Voicelab__trurl-2-13b' }
{ count: 239,
  dataset: 'open-llm-leaderboard/details_huggyllama__llama-65b' }
Type "it" for more
it
{ count: 239,
  dataset: 'open-llm-leaderboard/details_togethercomputer__RedPajama-INCITE-Base-3B-v1' }
{ count: 239, dataset: 'codeparrot/github-code-clean' }
{ count: 239,
  dataset: 'open-llm-leaderboard/details_Writer__palmyra-base' }
{ count: 239,
  dataset: 'open-llm-leaderboard/details_facebook__galactica-30b' }
{ count: 238,
  dataset: 'open-llm-leaderboard/details_acrastt__Marx-3B' }
{ count: 191,
  dataset: 'open-llm-leaderboard/details_voidful__changpt-bart' }
{ count: 191,
  dataset: 'open-llm-leaderboard/details_Charlie911__vicuna-7b-v1.5-lora-mctaco' }
{ count: 191,
  dataset: 'open-llm-leaderboard/details_AlekseyKorshuk__pygmalion-6b-vicuna-chatml' }
{ count: 191,
  dataset: 'open-llm-leaderboard/details_klosax__pythia-70m-deduped-step44k-92bt' }
{ count: 190,
  dataset: 'open-llm-leaderboard/details_Open-Orca__OpenOrca-Preview1-13B' }
{ count: 186,
  dataset: 'open-llm-leaderboard/details_TigerResearch__tigerbot-70b-chat' }
{ count: 186,
  dataset: 'open-llm-leaderboard/details_openBuddy__openbuddy-llama2-34b-v11.1-bf16' }
{ count: 186,
  dataset: 'open-llm-leaderboard/details_dhmeltzer__llama-7b-SFT_eli5_wiki65k_1024_r_64_alpha_16' }
{ count: 186,
  dataset: 'open-llm-leaderboard/details_uukuguy__speechless-orca-platypus-coig-lite-4k-0.6e-13b' }
...
</code></pre>
</details>

**Comment by severo** [comment]
Hmmm: the stats are the same as yesterday, and the datasets have no pending jobs. I'll try to restart the jobs. There seems to be an issue due to the config or split names, which surely makes the entries obsolete.

@AndreaFrancis could you give me a hand with that?

**Comment by severo** [comment]
For `dabdib/collections`, the `dataset-config-names` step now gives an error, which means that all the config and split level cache entries should be deleted (logically, but it's not implemented).

**Comment by severo** [comment]
Could be solved with https://github.com/huggingface/datasets-server/issues/1823#issuecomment-1736284469.

**Comment by severo** [comment]
Let's use the new admin endpoint (https://github.com/huggingface/datasets-server/pull/1884) once deployed to refresh all these datasets

**Comment by severo** [comment]
> Let's use the new admin endpoint (https://github.com/huggingface/datasets-server/pull/1884) once deployed to refresh all these datasets

I used it (and also deleted everything from the `open-llm-leaderboard` org, which is blocked). Now: we don't have any outdated dataset.
